protocol/proto/conn_enc_dec.go

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
```

Comment:
Feels like this `ConnectionEncoderDecoder` interface isn't all that useful. Seems like all it does atm is just delegating the encoding/decoding logic to the actual encoder/decoder. Might be cleaner to merge what's in here to the `AddressEncoderDecoder` to reduce cognitive burden.

Reply:
It allows resetting connection so it can be reused and pooled which is ideal for the aggregator consumer use case, it does not need the logic to retry tcp dials, which is much simpler. 
While AddressEncodeDecoder have the extra logic to retry connections which is only needed for producer cases. Originally I had them together but felt a separation is much cleaner.